shared_client.go

```diff
@@ -143,21 +144,55 @@ func handler(wg *sync.WaitGroup, client *Client, conn *Conn, requests chan reque
 
 	responses := make(chan sharedClientResponse)
 
+	// receiverTrigger is used to wake up the receive loop after request(s) have been sent. It
+	// must be buffered to be able to send a trigger while the receive loop is not yet ready to
+	// receive the trigger, as we do not want to stall the sender when the receiver is blocking
+	// on the read operation.
+	receiverTrigger := make(chan struct{}, 1)
+	triggerReceiver := func() {
+		select {
+		case receiverTrigger <- struct{}{}:
+		default:
+		}
+	}
+
 	// Receive loop
 	wg.Add(1)
 	go func() {
 		defer wg.Done()
 		defer close(responses)
+
+		// No point trying to receive until the first request has been successfully sent, so
+		// wait for a trigger first. receiverTrigger is buffered, so this is safe
+		// to do, even if the sender sends the trigger before we are ready to receive here.
+		<-receiverTrigger
```

Comment:
Don't the blocking semantics of `ReadMsg` take care of this anyway?

Reply:
Yes, but I don't know the details on what happens to a currently blocking ReadMsg when teh sending side is setting the read/write deadlines before each send operation. Likely it adjusts the timeouts, but could also wait forever based on the read timeout that was in effect when the blocking started?

Reply:
Oh, no the deadline will apply:

from `net.Conn` interface docs:

```
	// A deadline is an absolute time after which I/O operations
	// fail instead of blocking. The deadline applies to all future
	// and pending I/O, not just the immediately following call to
	// Read or Write. After a deadline has been exceeded, the
	// connection can be refreshed by setting a deadline in the future.
```

Reply:
Thanks for digging this up.